#!/usr/bin/env cwl-runner

cwlVersion: v1.0
class: CommandLineTool


#hints:
#- $import: envvar-global.yml
#- $import: GATK-docker.yml

inputs: # position 0, for java args, 1 for the jar, 2 for the tool itself

  GATKJar:
    type: File
    inputBinding:
      position: 1
      prefix: -jar
  multithreading_nt:
    type: int
    default: 4
    inputBinding:
      position: 2
      prefix: --nt
    doc: multithreading option

  reference:
    type: File
    inputBinding:
      position: 2
      prefix: -R
    doc: reference genome

  recal_file:
    type: File
    inputBinding:
      position: 2
      prefix: -recalFile
    doc: the recal file generated by VariantRecalibrator

  tranches_file:
    type: File
    inputBinding:
      position: 2
      prefix: -tranchesFile
    doc: the tranches file generated by VariantRecalibrator

  mode:
    type: string
    inputBinding:
      position: 2
      prefix: -mode
    doc: specify if VQSR is called on SNPs or Indels

  ts_filter_level:
    type: float
    default: 99.9
    inputBinding:
      position: 2
      prefix: --ts_filter_level
    doc: filtering level default value is 99.9

  java_mem:
    type: string
    default: -Xmx8g
    inputBinding:
      position: 0

outputs:
  vqsr_vcf:
    type: File
    outputBinding:
      glob: vqsr.vcf
    doc: The output recalibration VCF file


arguments:
- valueFrom: $(runtime.tmpdir)
  position: 0
  separate: false
  prefix: -Djava.io.tmpdir=
- valueFrom: ApplyRecalibration
  position: 2
  prefix: -T
- valueFrom: vqsr.vcf
  position: 2
  prefix: -o
baseCommand: [java]

